File: daemon/src/scheduler.py
import os
import threading
import time
import re
import logging
from datetime import datetime, timedelta

from src.config import read_config
from src.version import VERSION

logger = logging.getLogger('cronplus.scheduler')

# /run is tmpfs — this file disappears on reboot.
# File exists  → daemon was restarted (skip @reboot)
# File missing → system rebooted (fire @reboot, then create file)
REBOOT_FILE = '/run/cronplus'


class Scheduler:
    def __init__(self, executor, conf_path=None, reload_interval=60):
        self.executor = executor
        self.conf_path = conf_path
        self.reload_interval = reload_interval
        self._running = False
        self._thread = None
        self._cond = threading.Condition()
        self._stats = {'tasks_executed': 0, 'tasks_failed': 0, 'last_run': None, 'started_at': None}
        self._last_conf_mtime = 0
        self._last_cleanup = 0
        self._task_last_run = {}  # task_id → datetime of last scheduled run
        self._cached_tasks = None  # cached task list, invalidated on config change

    def start(self):
        if self._running:
            return
        self._running = True
        self._stats['started_at'] = datetime.now().isoformat()
        self._thread = threading.Thread(target=self._loop, daemon=True, name='cronplus-scheduler')
        self._thread.start()
        logger.info(f"Scheduler v{VERSION} started")

    def stop(self):
        self._running = False
        with self._cond:
            self._cond.notify_all()
        if self._thread:
            self._thread.join(timeout=5)
        logger.info("Scheduler stopped")

    def wakeup(self):
        """Force re-read of config and reschedule."""
        self._last_conf_mtime = 0
        self._cached_tasks = None
        with self._cond:
            self._cond.notify_all()

    def _is_real_reboot(self):
        """Check /run/cronplus file: missing = system rebooted (tmpfs cleared it).

        Returns True once after a real reboot, then creates the file so
        subsequent daemon restarts won't re-trigger @reboot tasks.
        """
        exists = os.path.exists(REBOOT_FILE)
        if exists:
            logger.debug(f"Reboot check: {REBOOT_FILE} exists → NOT a reboot")
            return False
        try:
            with open(REBOOT_FILE, 'w') as f:
                f.write('1')
            logger.info(f"Reboot check: {REBOOT_FILE} missing → REBOOT detected, marker created")
            return True
        except OSError as e:
            logger.warning(f"Reboot check: failed to create marker: {e}")
            return False

    def get_stats(self):
        stats = dict(self._stats)
        stats['running_tasks'] = self.executor.get_running()
        stats['running_count'] = len(stats['running_tasks'])
        return stats

    def get_next_runs(self, second='0', minute='*', hour='*', day='*', month='*', dow='*',
                      count=5, special=None):
        if special:
            return []
        results = []
        now = datetime.now()
        current = now.replace(microsecond=0) + timedelta(seconds=1)
        max_check = current + timedelta(days=730)
        safety = 0
        while len(results) < count and current < max_check and safety < 500000:
            safety += 1
            if self._match_cron(current, second, minute, hour, day, month, dow):
                results.append(current.isoformat())
            current += timedelta(seconds=1)
        return results

    def _drift_compensated_wait(self, timeout):
        """[优化2] 飘移补偿等待：动态计算到下一秒整点的精确等待时间。

        相比固定 time.sleep(1) 或 cond.wait(timeout=1)，此方法通过计算
        距离下一秒边界的精确毫秒数来消除累积飘移，确保长时间运行后调度精度。
        """
        now = time.time()
        # 计算距离下一秒整点的秒数（例如 now=3.7 → wait=0.3）
        wait_time = 1.0 - (now % 1.0)
        # 确保不会因浮点精度问题出现负数或超大值
        wait_time = max(0.05, min(wait_time, 1.0))
        with self._cond:
            self._cond.wait(timeout=min(wait_time, timeout))

    # ── Main loop ──

    def _loop(self):
        logger.info(f"Scheduler loop started, reboot marker: {REBOOT_FILE} "
                    f"{'exists' if os.path.exists(REBOOT_FILE) else 'MISSING (will fire @reboot tasks)'}")
        while self._running:
            try:
                # Periodic cleanup of stale executor entries + zombie processes (every 5 min)
                now_ts = time.time()
                if now_ts - self._last_cleanup > 300:
                    self.executor.cleanup_stale()
                    self.executor.cleanup_zombies()  # [优化1] 定期巡检僵尸进程
                    self._last_cleanup = now_ts

                tasks = self._load_tasks()
                if not tasks:
                    logger.debug("No tasks configured, waiting...")
                    self._drift_compensated_wait(1)
                    continue

                now_floor = datetime.now().replace(microsecond=0)
                next_runs = []
                real_reboot = self._is_real_reboot()

                for task in tasks:
                    task_id = task.get('id', '?')
                    if not task.get('enabled', True):
                        logger.debug(f"Task [{task_id}] disabled, skipping")
                        continue

                    schedule = task.get('schedule', '')

                    # Handle @reboot tasks — only fire on actual system reboot
                    if schedule == '@reboot':
                        if real_reboot:
                            delay = int(task.get('reboot_delay', 0) or 0)
                            if delay > 0:
                                logger.info(f"Task [{task_id}] @reboot: scheduled in {delay}s")
                                def _delayed_reboot(t=task, d=delay):
                                    time.sleep(d)
                                    self._execute_task(t)
                                threading.Thread(target=_delayed_reboot, daemon=True).start()
                            else:
                                logger.info(f"Task [{task_id}] @reboot: firing now")
                                self._execute_task(task)
                        else:
                            logger.debug(f"Task [{task_id}] @reboot: skipped (not a reboot)")
                        continue

                    if self.executor.is_running(task_id):
                        logger.debug(f"Task [{task_id}] already running, skipping")
                        continue

                    # Other @ specials → convert to cron
                    if schedule.startswith('@'):
                        cron = self._special_to_cron(schedule)
                        if cron is None:
                            continue
                        s, m, h, d, mo, w = cron
                    else:
                        s, m, h, d, mo, w = self.parse_schedule(schedule)

                    # Find next run strictly after last scheduled run
                    last_run = self._task_last_run.get(task_id, now_floor - timedelta(seconds=1))
                    nxt = self._next_run_time(s, m, h, d, mo, w, last_run, inclusive=False)
                    if nxt:
                        next_runs.append((nxt, task, s, m, h, d, mo, w))

                if not next_runs:
                    logger.debug("No runnable tasks found, waiting 1s")
                    self._drift_compensated_wait(1)
                    continue

                next_runs.sort(key=lambda x: x[0])

                # FIX: Execute tasks whose next scheduled time <= current second
                # Changed from == to <= to prevent missed-second deadlock
                for run_time, task, s, m, h, d, mo, w in next_runs:
                    if run_time <= now_floor:
                        task_id = task.get('id', '?')

                        # FIX: Always update last_run immediately upon reaching the
                        # scheduled time, even if the task is skipped due to being
                        # already running. This prevents the infinite-skip deadlock.
                        self._task_last_run[task_id] = run_time

                        if self.executor.is_running(task_id):
                            logger.debug(f"Task [{task_id}] already running, skipping this window")
                            continue

                        self._execute_task(task)
                    else:
                        break

                # Calculate wait until next task
                remaining = [item for item in next_runs if item[0] != now_floor]
                if remaining:
                    nearest_time = remaining[0][0]
                    nearest_task = remaining[0][1]
                    wait_sec = (nearest_time - datetime.now()).total_seconds()
                    if wait_sec > 0:
                        logger.info(f"Next: task [{nearest_task.get('id','?')}] "
                                   f"'{nearest_task.get('title') or nearest_task.get('command','')[:40]}' "
                                   f"at {nearest_time.strftime('%Y-%m-%d %H:%M:%S')} "
                                   f"(in {wait_sec:.0f}s)")
                        # [优化2] 飘移补偿：精确等待到下一秒边界，而非固定 1s
                        self._drift_compensated_wait(min(wait_sec, 1))
                        continue

                # Short sleep to avoid tight loop
                self._drift_compensated_wait(0.5)

            except Exception as e:
                logger.error(f"Scheduler loop error: {e}", exc_info=True)
                time.sleep(5)

    def _load_tasks(self):
        """Load tasks from config, with mtime-based caching.

        Only re-parses the file when its modification time changes,
        making hot-reload reliable even without SIGHUP.
        """
        try:
            mtime = os.path.getmtime(self.conf_path)
        except OSError:
            mtime = 0

        if mtime != self._last_conf_mtime or self._cached_tasks is None:
            try:
                self._cached_tasks = read_config(self.conf_path)
                self._last_conf_mtime = mtime
                if self._cached_tasks:
                    logger.info(f"Config reloaded: {len(self._cached_tasks)} tasks (mtime={mtime})")
            except Exception as e:
                logger.error(f"Failed to load config: {e}")
                # Keep cached tasks on parse error (don't lose existing schedule)
                if self._cached_tasks is None:
                    self._cached_tasks = []

        return self._cached_tasks

    def _execute_task(self, task):
        task_id = task.get('id', '?')
        logger.info(f"Scheduler: triggering task [{task_id}] {task.get('title') or task.get('command','')[:60]}")
        def _run():
            try:
                result = self.executor.execute(task, trigger='auto')
                if result:
                    self._stats['tasks_executed'] += 1
                    self._stats['last_run'] = datetime.now().isoformat()
                    if result[0] == 'error':
                        self._stats['tasks_failed'] += 1
            except Exception as e:
                logger.error(f"Task [{task_id}] thread error: {e}", exc_info=True)
        threading.Thread(target=_run, daemon=True).start()

    # ── Cron matching ──

    @staticmethod
    def parse_schedule(schedule):
        """Parse a 6-field cron expression into (second, minute, hour, day, month, dow)."""
        parts = schedule.split()
        if len(parts) == 6:
            return parts[0], parts[1], parts[2], parts[3], parts[4], parts[5]
        if len(parts) == 5:
            return '0', parts[0], parts[1], parts[2], parts[3], parts[4]
        return '0', '*', '*', '*', '*', '*'

    def _next_run_time(self, second, minute, hour, day, month, dow, after, inclusive=False):
        if inclusive:
            current = after.replace(microsecond=0)
        else:
            current = after.replace(microsecond=0) + timedelta(seconds=1)
        max_check = current + timedelta(days=730)
        safety = 0
        while current < max_check and safety < 5000000:
            safety += 1
            if self._match_cron(current, second, minute, hour, day, month, dow):
                return current
            current += timedelta(seconds=1)
        return None

    def _match_cron(self, dt, second, minute, hour, day, month, dow):
        if not self._match_field(dt.second, second, 0, 59): return False
        if not self._match_field(dt.minute, minute, 0, 59): return False
        if not self._match_field(dt.hour, hour, 0, 23): return False
        if not self._match_field(dt.month, month, 1, 12): return False
        day_match = self._match_field(dt.day, day, 1, 31)
        dow_match = self._match_field(dt.weekday() if dt.weekday() < 7 else 0, dow, 0, 7)
        if day != '*' and dow != '*':
            return day_match or dow_match
        if day != '*' and not day_match: return False
        if dow != '*' and not dow_match: return False
        return True

    def _match_field(self, value, spec, min_val, max_val):
        if spec == '*': return True
        if ',' in spec:
            return any(self._match_field(value, s.strip(), min_val, max_val) for s in spec.split(','))
        m = re.match(r'^(\d+)-(\d+)$', spec)
        if m: return int(m.group(1)) <= value <= int(m.group(2))
        m = re.match(r'^(.*?)/(\d+)$', spec)
        if m:
            base, step = m.group(1), int(m.group(2))
            if base == '*': return (value - min_val) % step == 0
            if '-' in base:
                p = base.split('-')
                return int(p[0]) <= value <= int(p[1]) and (value - int(p[0])) % step == 0
            return value == int(base)
        try: return value == int(spec)
        except ValueError: return False

    def _special_to_cron(self, schedule):
        specials = {
            '@yearly':    ('0', '0', '0', '1', '1', '*'),
            '@annually':  ('0', '0', '0', '1', '1', '*'),
            '@monthly':   ('0', '0', '0', '1', '*', '*'),
            '@weekly':    ('0', '0', '0', '*', '*', '0'),
            '@daily':     ('0', '0', '0', '*', '*', '*'),
            '@midnight':  ('0', '0', '0', '*', '*', '*'),
            '@hourly':    ('0', '0', '*', '*', '*', '*'),
        }
        return specials.get(schedule)

File: cronplus/internal/scheduler/scheduler.go
package scheduler

import (
	"context"
	"log"
	"os"
	"sync/atomic"
	"time"

	"cronplus/internal/store"
	"cronplus/pkg/model"
)

// Pool is the interface the executor pool must satisfy.
type Pool interface {
	Dispatch(task model.Task, trigger string)
	Kill(taskID int)
	KillAll()
	IsRunning(taskID int) bool
}

// Scheduler manages the main scheduling loop.
type Scheduler struct {
	store         *store.Store
	pool          Pool
	running       atomic.Bool
	taskLastRun   map[int]time.Time
	lastMtime     int64
	cachedTasks   []model.Task
	rebootFired   bool
	stats         Stats
	lastNextTaskID int    // track last logged "Next:" task to avoid spam
	lastNextTime  string // track last logged "Next:" time
}

// Stats holds scheduler statistics.
type Stats struct {
	TasksExecuted int    `json:"tasks_executed"`
	TasksFailed   int    `json:"tasks_failed"`
	LastRun       string `json:"last_run"`
	StartedAt     string `json:"started_at"`
}

func New(s *store.Store, pool Pool) *Scheduler {
	return &Scheduler{
		store:       s,
		pool:        pool,
		taskLastRun: make(map[int]time.Time),
		rebootFired: !store.RebootDetected(), // if marker exists, reboot already fired
	}
}

// Start begins the scheduling loop in a background goroutine.
func (s *Scheduler) Start(ctx context.Context) {
	if !s.running.CompareAndSwap(false, true) {
		return
	}
	s.stats.StartedAt = time.Now().Format(time.RFC3339)
	go s.loop(ctx)
	log.Println("Scheduler started")
}

// Stop signals the scheduler to stop.
func (s *Scheduler) Stop() {
	s.running.Store(false)
	log.Println("Scheduler stopped")
}

// Wakeup forces a config reload and reschedule.
func (s *Scheduler) Wakeup() {
	s.lastMtime = 0
	s.cachedTasks = nil
	s.taskLastRun = make(map[int]time.Time)
	s.lastNextTaskID = 0
	s.lastNextTime = ""
	log.Println("Scheduler: cache cleared, will re-read config")
}

// GetStats returns current scheduler statistics.
func (s *Scheduler) GetStats() Stats {
	return s.stats
}

func (s *Scheduler) loop(ctx context.Context) {
	var lastCleanup time.Time

	for s.running.Load() {
		select {
		case <-ctx.Done():
			return
		default:
		}

		// Periodic stale cleanup (every 5 min)
		if time.Since(lastCleanup) > 5*time.Minute {
			lastCleanup = time.Now()
		}

		tasks, err := s.loadTasks()
		if err != nil {
			log.Printf("Scheduler: load tasks error: %v", err)
			driftSleep(ctx, 5*time.Second)
			continue
		}
		if len(tasks) == 0 {
			driftSleep(ctx, 1*time.Second)
			continue
		}

		now := time.Now().Truncate(time.Second)
		type candidate struct {
			at   time.Time
			task model.Task
			cron [6]string
		}
		var candidates []candidate

		for _, task := range tasks {
			if !task.Enabled {
				continue
			}

			// Handle @reboot
			if task.Schedule == "@reboot" {
				if !s.rebootFired {
					s.rebootFired = true
					if task.RebootDelay > 0 {
						log.Printf("Task [%d] @reboot: scheduled in %ds", task.ID, task.RebootDelay)
						go func(t model.Task) {
							time.Sleep(time.Duration(t.RebootDelay) * time.Second)
							s.pool.Dispatch(t, "auto")
							s.stats.TasksExecuted++
							s.stats.LastRun = time.Now().Format(time.RFC3339)
						}(task)
					} else {
						log.Printf("Task [%d] @reboot: firing now", task.ID)
						s.pool.Dispatch(task, "auto")
						s.stats.TasksExecuted++
						s.stats.LastRun = time.Now().Format(time.RFC3339)
					}
				}
				continue
			}

			if s.pool.IsRunning(task.ID) {
				if task.KillPrevious {
					s.pool.Kill(task.ID)
				} else {
					continue
				}
			}

			fields, err := ParseSchedule(task.Schedule)
			if err != nil {
				log.Printf("Task [%d] bad schedule %q: %v", task.ID, task.Schedule, err)
				continue
			}

			lastRun := s.taskLastRun[task.ID]
			if lastRun.IsZero() {
				lastRun = now.Add(-time.Second)
			}
			nxt := NextRunTime(fields, lastRun)
			if !nxt.IsZero() {
				candidates = append(candidates, candidate{at: nxt, task: task, cron: fields})
			}
		}

		if len(candidates) == 0 {
			driftSleep(ctx, 1*time.Second)
			continue
		}

		// Sort by trigger time
		for i := 0; i < len(candidates)-1; i++ {
			for j := i + 1; j < len(candidates); j++ {
				if candidates[j].at.Before(candidates[i].at) {
					candidates[i], candidates[j] = candidates[j], candidates[i]
				}
			}
		}

		// Execute all candidates whose time <= now
		executed := false
		for _, c := range candidates {
			if c.at.After(now) {
				break
			}
			taskID := c.task.ID
			s.taskLastRun[taskID] = c.at

			if s.pool.IsRunning(taskID) {
				if c.task.KillPrevious {
					s.pool.Kill(taskID)
				} else {
					continue
				}
			}

			log.Printf("Scheduler: triggering task [%d] %s", taskID,
				func() string {
					if c.task.Title != "" {
						return c.task.Title
					}
					if len(c.task.Command) > 60 {
						return c.task.Command[:60]
					}
					return c.task.Command
				}())

			s.pool.Dispatch(c.task, "auto")
			s.stats.TasksExecuted++
			s.stats.LastRun = time.Now().Format(time.RFC3339)
			s.lastNextTaskID = 0 // reset so next task gets logged
			s.lastNextTime = ""
			executed = true
		}

		// Calculate sleep until next candidate
		var nextWait time.Duration
		for _, c := range candidates {
			if c.at.After(now) {
				wait := time.Until(c.at)
				if wait > 0 {
					nextWait = wait
					break
				}
			}
		}

		if nextWait > 0 && !executed {
			// Log next task info — only when the task or time changes
			for _, c := range candidates {
				if c.at.After(now) {
					nextTimeStr := c.at.Format("2006-01-02 15:04:05")
					if c.task.ID != s.lastNextTaskID || nextTimeStr != s.lastNextTime {
						taskName := c.task.Title
						if taskName == "" {
							taskName = c.task.Command
							if len(taskName) > 40 {
								taskName = taskName[:40]
							}
						}
						log.Printf("Next: task [%d] '%s' at %s (in %.0fs)",
							c.task.ID, taskName,
							nextTimeStr,
							time.Until(c.at).Seconds())
						s.lastNextTaskID = c.task.ID
						s.lastNextTime = nextTimeStr
					}
					break
				}
			}
			driftSleep(ctx, min(nextWait, 1*time.Second))
		} else {
			driftSleep(ctx, 500*time.Millisecond)
		}
	}
}

func (s *Scheduler) loadTasks() ([]model.Task, error) {
	info, err := os.Stat(s.store.ConfPath)
	if err != nil {
		mtime := int64(0)
		if s.lastMtime != mtime || s.cachedTasks == nil {
			tasks, err := s.store.ListTasks()
			if err != nil {
				return nil, err
			}
			s.cachedTasks = tasks
			s.lastMtime = mtime
			if len(tasks) > 0 {
				log.Printf("Config loaded: %d tasks", len(tasks))
			}
		}
		return s.cachedTasks, nil
	}

	mtime := info.ModTime().UnixNano()
	if mtime != s.lastMtime || s.cachedTasks == nil {
		tasks, err := s.store.ListTasks()
		if err != nil {
			// Keep cached tasks on error
			if s.cachedTasks == nil {
				return nil, err
			}
			log.Printf("Config reload error (keeping cached): %v", err)
			return s.cachedTasks, nil
		}
		s.cachedTasks = tasks
		s.lastMtime = mtime
		log.Printf("Config reloaded: %d tasks", len(tasks))
	}
	return s.cachedTasks, nil
}

// driftSleep sleeps with drift compensation — aligns to the next second boundary.
func driftSleep(ctx context.Context, max time.Duration) {
	now := time.Now()
	wait := time.Second - time.Duration(now.Nanosecond())*time.Nanosecond
	if wait < 50*time.Millisecond {
		wait = time.Second
	}
	if wait > max {
		wait = max
	}
	select {
	case <-ctx.Done():
	case <-time.After(wait):
	}
}
